Add sale_condition_template

guewen opened the pull request:
I moved this module from sale-workflow.  Displaying additional condition texts on sales reports is clearly a reporting module.

This module has been migrated from sale_condition_text, here is the original PR for the migration before I moved the module: https://github.com/OCA/sale-workflow/pull/50

This is still a WIP module because the report does not show the fields at the moment.

Depends on :
- [x] https://github.com/OCA/account-invoice-reporting/pull/14
- [x] https://github.com/OCA/account-invoice-reporting/pull/15

nbessi committed:
[ADD] sale_condition_text
(lp:c2c-addons/6.1  rev 19)

nbessi committed:
[ADD] module description
(lp:c2c-addons/6.1  rev 20)

nbessi committed:
[MRG] 6.1
(lp:c2c-addons/6.1  rev 28.1.5)

Commit:
[MIGR] set installable to False on all the modules, the flag has to be reactivated when the modules are migrated

Conflicts:
	__unported__/html_sale_product_note/__openerp__.py
	__unported__/sale_note_flow/__openerp__.py
	mail_quotation/__openerp__.py
	pricelist_share_companies/__openerp__.py
	product_special_type/__openerp__.py

eLBati committed:
[FIX] repository PEP8

Conflicts:
	__unported__/html_sale_product_note/__init__.py
	__unported__/html_sale_product_note/product.py
	__unported__/sale_note_flow/__init__.py
	__unported__/sale_note_flow/sale.py
	__unported__/sale_order_proforma_webkit/__init__.py
	__unported__/sale_order_webkit/__init__.py
	__unported__/sale_order_webkit/__openerp__.py
	account_invoice_reorder_lines/__init__.py
	account_invoice_reorder_lines/__openerp__.py
	account_invoice_reorder_lines/invoice.py
	mail_quotation/__init__.py
	mail_quotation/__openerp__.py
	partner_prepayment/model/res_partner.py
	partner_prepayment/model/sale.py
	pricelist_share_companies/__init__.py
	pricelist_share_companies/__openerp__.py
	pricelist_share_companies/company.py
	pricelist_share_companies/pricelist.py
	product_customer_code_sale/__init__.py
	product_customer_code_sale/__openerp__.py
	product_customer_code_sale/sale.py
	product_special_type/__openerp__.py
	product_special_type/product.py
	product_special_type_invoice/__openerp__.py
	product_special_type_invoice/invoice.py
	product_special_type_sale/sale.py
	purchase_order_reorder_lines/__init__.py
	purchase_order_reorder_lines/__openerp__.py
	purchase_order_reorder_lines/purchase.py
	sale_cancel_reason/__init__.py
	sale_cancel_reason/model/sale.py
	sale_cancel_reason/wizard/__init__.py
	sale_cancel_reason/wizard/cancel_reason.py
	sale_delivery_term/__init__.py
	sale_delivery_term/__openerp__.py
	sale_delivery_term/sale.py
	sale_dropshipping/__init__.py
	sale_dropshipping/__openerp__.py
	sale_dropshipping/product.py
	sale_dropshipping/purchase.py
	sale_dropshipping/sale.py
	sale_exception_nostock/__init__.py
	sale_exception_nostock/__openerp__.py
	sale_exception_nostock/model/__init__.py
	sale_exception_nostock/model/sale.py
	sale_exception_nostock/test/__init__.py
	sale_exceptions/__init__.py
	sale_exceptions/__openerp__.py
	sale_exceptions/sale.py
	sale_exceptions/wizard/__init__.py
	sale_exceptions/wizard/sale_exception_confirm.py
	sale_fiscal_position_update/__init__.py
	sale_fiscal_position_update/__openerp__.py
	sale_fiscal_position_update/sale.py
	sale_jit_on_services/__init__.py
	sale_jit_on_services/__openerp__.py
	sale_jit_on_services/procurement.py
	sale_journal_shop/__init__.py
	sale_journal_shop/__openerp__.py
	sale_journal_shop/sale.py
	sale_multi_picking/__init__.py
	sale_multi_picking/__openerp__.py
	sale_multi_picking/sale.py
	sale_order_force_number/__init__.py
	sale_order_force_number/__openerp__.py
	sale_quotation_number/__init__.py
	sale_quotation_number/__openerp__.py
	sale_quotation_number/sale_quotation.py
	sale_sourced_by_line/__init__.py
	sale_sourced_by_line/model/__init__.py
	sale_sourced_by_line/model/sale.py
	sale_stock_global_delivery_lead_time/__init__.py
	sale_stock_global_delivery_lead_time/__openerp__.py
	sale_stock_global_delivery_lead_time/sale_stock.py
	sale_validity/__init__.py
	sale_validity/model/company.py
	sale_validity/model/sale_order.py
	stock_picking_reorder_lines/__init__.py
	stock_picking_reorder_lines/__openerp__.py
	stock_picking_reorder_lines/stock.py

eLBati committed:
[FIX] Still some PEP8 erros

Conflicts:
	sale_cancel_reason/model/sale.py
	sale_dropshipping/purchase.py
	sale_quotation_number/sale_quotation.py

pedrobaeza committed:
- Set as uninstallable and moved to __unported__ dir all modules.
- Update travis.yml to 8.0

Conflicts:
	.travis.yml
	__unported__/account_invoice_reorder_lines/__init__.py
	__unported__/account_invoice_reorder_lines/__openerp__.py
	__unported__/account_invoice_reorder_lines/i18n/account_invoice_reorder_lines.pot
	__unported__/account_invoice_reorder_lines/i18n/en.po
	__unported__/account_invoice_reorder_lines/i18n/fr.po
	__unported__/account_invoice_reorder_lines/invoice.py
	__unported__/account_invoice_reorder_lines/invoice_view.xml
	__unported__/html_sale_product_note/__init__.py
	__unported__/html_sale_product_note/__openerp__.py
	__unported__/html_sale_product_note/product.py
	__unported__/mail_quotation/__init__.py
	__unported__/mail_quotation/__openerp__.py
	__unported__/mail_quotation/quotation_action_data.xml
	__unported__/mail_quotation/sale_order_view.xml
	__unported__/partner_prepayment/__init__.py
	__unported__/partner_prepayment/__openerp__.py
	__unported__/partner_prepayment/model/__init__.py
	__unported__/partner_prepayment/model/res_partner.py
	__unported__/partner_prepayment/model/sale.py
	__unported__/partner_prepayment/test/sale_order_prepaid.yml
	__unported__/partner_prepayment/view/partner_view.xml
	__unported__/pricelist_share_companies/__init__.py
	__unported__/pricelist_share_companies/__openerp__.py
	__unported__/pricelist_share_companies/company.py
	__unported__/pricelist_share_companies/company_view.xml
	__unported__/pricelist_share_companies/pricelist.py
	__unported__/pricelist_share_companies/pricelist_view.xml
	__unported__/pricelist_share_companies/security/pricelist_security.xml
	__unported__/product_customer_code_sale/__init__.py
	__unported__/product_customer_code_sale/__openerp__.py
	__unported__/product_customer_code_sale/i18n/it.po
	__unported__/product_customer_code_sale/i18n/product_customer_code_sale.pot
	__unported__/product_customer_code_sale/sale.py
	__unported__/product_customer_code_sale/sale_view.xml
	__unported__/product_special_type/__init__.py
	__unported__/product_special_type/__openerp__.py
	__unported__/product_special_type/product.py
	__unported__/product_special_type/product_view.xml
	__unported__/product_special_type_invoice/__init__.py
	__unported__/product_special_type_invoice/__openerp__.py
	__unported__/product_special_type_invoice/invoice.py
	__unported__/product_special_type_sale/__init__.py
	__unported__/product_special_type_sale/__openerp__.py
	__unported__/product_special_type_sale/sale.py
	__unported__/purchase_order_reorder_lines/__init__.py
	__unported__/purchase_order_reorder_lines/__openerp__.py
	__unported__/purchase_order_reorder_lines/i18n/en.po
	__unported__/purchase_order_reorder_lines/i18n/fr.po
	__unported__/purchase_order_reorder_lines/purchase.py
	__unported__/purchase_order_reorder_lines/purchase_view.xml
	__unported__/sale_cancel_reason/__init__.py
	__unported__/sale_cancel_reason/__openerp__.py
	__unported__/sale_cancel_reason/data/sale_order_cancel_reason.xml
	__unported__/sale_cancel_reason/model/__init__.py
	__unported__/sale_cancel_reason/model/sale.py
	__unported__/sale_cancel_reason/security/ir.model.access.csv
	__unported__/sale_cancel_reason/test/sale_order_cancel.yml
	__unported__/sale_cancel_reason/view/sale_view.xml
	__unported__/sale_cancel_reason/wizard/__init__.py
	__unported__/sale_cancel_reason/wizard/cancel_reason.py
	__unported__/sale_cancel_reason/wizard/cancel_reason_view.xml
	__unported__/sale_delivery_term/AUTHORS.txt
	__unported__/sale_delivery_term/__init__.py
	__unported__/sale_delivery_term/__openerp__.py
	__unported__/sale_delivery_term/sale.py
	__unported__/sale_delivery_term/sale_data.xml
	__unported__/sale_delivery_term/sale_demo.xml
	__unported__/sale_delivery_term/sale_view.xml
	__unported__/sale_delivery_term/security/ir.model.access.csv
	__unported__/sale_delivery_term/test/sale_order_demo.yml
	__unported__/sale_dropshipping/__init__.py
	__unported__/sale_dropshipping/__openerp__.py
	__unported__/sale_dropshipping/i18n/pt_BR.po
	__unported__/sale_dropshipping/images/purchase_to_sale.png
	__unported__/sale_dropshipping/product.py
	__unported__/sale_dropshipping/product_view.xml
	__unported__/sale_dropshipping/purchase.py
	__unported__/sale_dropshipping/purchase_view.xml
	__unported__/sale_dropshipping/sale.py
	__unported__/sale_dropshipping/sale_view.xml
	__unported__/sale_dropshipping/stock_view.xml
	__unported__/sale_dropshipping/test/test_sale_policy_procurement.yml
	__unported__/sale_exception_nostock/__init__.py
	__unported__/sale_exception_nostock/__openerp__.py
	__unported__/sale_exception_nostock/data/data.xml
	__unported__/sale_exception_nostock/model/__init__.py
	__unported__/sale_exception_nostock/model/sale.py
	__unported__/sale_exception_nostock/test/__init__.py
	__unported__/sale_exception_nostock/test/no_stock_test.yml
	__unported__/sale_exception_nostock/test/test_utils.py
	__unported__/sale_exception_nostock_by_line/__init__.py
	__unported__/sale_exception_nostock_by_line/__openerp__.py
	__unported__/sale_exception_nostock_by_line/model/__init__.py
	__unported__/sale_exception_nostock_by_line/model/sale.py
	__unported__/sale_exceptions/__init__.py
	__unported__/sale_exceptions/__openerp__.py
	__unported__/sale_exceptions/i18n/es.po
	__unported__/sale_exceptions/i18n/fr.po
	__unported__/sale_exceptions/i18n/sale_exceptions.pot
	__unported__/sale_exceptions/sale.py
	__unported__/sale_exceptions/sale_exceptions_data.xml
	__unported__/sale_exceptions/sale_view.xml
	__unported__/sale_exceptions/sale_workflow.xml
	__unported__/sale_exceptions/security/ir.model.access.csv
	__unported__/sale_exceptions/settings/sale.exception.csv
	__unported__/sale_exceptions/wizard/__init__.py
	__unported__/sale_exceptions/wizard/sale_exception_confirm.py
	__unported__/sale_exceptions/wizard/sale_exception_confirm_view.xml
	__unported__/sale_fiscal_position_update/__init__.py
	__unported__/sale_fiscal_position_update/__openerp__.py
	__unported__/sale_fiscal_position_update/i18n/fr.po
	__unported__/sale_fiscal_position_update/i18n/sale_fiscal_position_update.pot
	__unported__/sale_fiscal_position_update/sale.py
	__unported__/sale_fiscal_position_update/sale_view.xml
	__unported__/sale_jit_on_services/__init__.py
	__unported__/sale_jit_on_services/__openerp__.py
	__unported__/sale_jit_on_services/procurement.py
	__unported__/sale_jit_on_services/procurement_workflow.xml
	__unported__/sale_jit_on_services/test/sale_service_jit_test.yml
	__unported__/sale_journal_shop/__init__.py
	__unported__/sale_journal_shop/__openerp__.py
	__unported__/sale_journal_shop/sale.py
	__unported__/sale_journal_shop/sale_view.xml
	__unported__/sale_line_description/__init__.py
	__unported__/sale_line_description/__openerp__.py
	__unported__/sale_line_description/i18n/fr.po
	__unported__/sale_line_description/i18n/it.po
	__unported__/sale_line_description/i18n/sale_line_description.pot
	__unported__/sale_line_description/res_config.py
	__unported__/sale_line_description/res_config_view.xml
	__unported__/sale_line_description/sale.py
	__unported__/sale_line_description/security/sale_security.xml
	__unported__/sale_multi_picking/AUTHORS.txt
	__unported__/sale_multi_picking/__init__.py
	__unported__/sale_multi_picking/__openerp__.py
	__unported__/sale_multi_picking/sale.py
	__unported__/sale_multi_picking/sale_demo.xml
	__unported__/sale_multi_picking/sale_view.xml
	__unported__/sale_multi_picking/security/ir.model.access.csv
	__unported__/sale_note_flow/__init__.py
	__unported__/sale_order_force_number/AUTHORS.txt
	__unported__/sale_order_force_number/__init__.py
	__unported__/sale_order_force_number/__openerp__.py
	__unported__/sale_order_force_number/sale_view.xml
	__unported__/sale_order_proforma_webkit/__init__.py
	__unported__/sale_order_webkit/__init__.py
	__unported__/sale_order_webkit/report/__init__.py
	__unported__/sale_quotation_number/AUTHORS.txt
	__unported__/sale_quotation_number/__init__.py
	__unported__/sale_quotation_number/__openerp__.py
	__unported__/sale_quotation_number/i18n/en_US.po
	__unported__/sale_quotation_number/i18n/zh_CN.po
	__unported__/sale_quotation_number/sale_quotation.py
	__unported__/sale_quotation_number/sale_quotation_view.xml
	__unported__/sale_sourced_by_line/__init__.py
	__unported__/sale_sourced_by_line/__openerp__.py
	__unported__/sale_sourced_by_line/model/__init__.py
	__unported__/sale_sourced_by_line/model/sale.py
	__unported__/sale_sourced_by_line/test/sale_order_not_sourced.yml
	__unported__/sale_sourced_by_line/test/sale_order_source.yml
	__unported__/sale_sourced_by_line/view/sale_view.xml
	__unported__/sale_stock_global_delivery_lead_time/__init__.py
	__unported__/sale_stock_global_delivery_lead_time/__openerp__.py
	__unported__/sale_stock_global_delivery_lead_time/sale_stock.py
	__unported__/sale_stock_global_delivery_lead_time/sale_stock_view.xml
	__unported__/sale_validity/__init__.py
	__unported__/sale_validity/__openerp__.py
	__unported__/sale_validity/i18n/fr.po
	__unported__/sale_validity/i18n/sale_validity.pot
	__unported__/sale_validity/model/__init__.py
	__unported__/sale_validity/model/company.py
	__unported__/sale_validity/model/sale_order.py
	__unported__/sale_validity/view/company_view.xml
	__unported__/sale_validity/view/sale_order.xml
	__unported__/stock_picking_reorder_lines/__init__.py
	__unported__/stock_picking_reorder_lines/__openerp__.py
	__unported__/stock_picking_reorder_lines/i18n/en.po
	__unported__/stock_picking_reorder_lines/i18n/fr.po
	__unported__/stock_picking_reorder_lines/stock.py
	__unported__/stock_picking_reorder_lines/stock_view.xml
	partner_prepayment/__init__.py
	partner_prepayment/model/__init__.py
	product_special_type/__init__.py
	sale_cancel_reason/__openerp__.py
	sale_cancel_reason/model/__init__.py
	sale_exception_nostock_by_line/model/sale.py
	sale_validity/model/__init__.py

yvaucher committed:
reactivate module sale_condition_text

yvaucher committed:
set proper license headers

yvaucher committed:
rename module to sale_condition_template

yvaucher committed:
improve manifest

yvaucher committed:
port sale_condition_template
* rename account.condition_text to base.condition_template
* move fields in view to reflect document print
* rename type of field to position
* give better position names

yvaucher committed:
split module sale_condition_template to extract base_condition_template

lepistone committed:
lint

Commit:
Remove base_condition_template from this repo, moved to
account-invoice-reporting

Conflicts:
	.travis.yml

Commit:
Remove the creation of invoice's fields, done in
invoice_condition_template

Commit:
The model 'base.condition_template' has been renamed to
'base.condition.template'

Commit:
Use an assert for an error that is concerning developers

Commit:
Extract the module description in README.rst

Commit:
The context is already used when calling write directly on a browse_record

Commit:
Remove blank line

Commit:
Add a menu to open the Document Conditions from the sale configuration
menu

Commit:
Make the API signature standard so the API wrapper can convert it

also use names for keyword arguments

Commit:
Add translations

Commit:
Added security accesses for the sales users and managers

guewen commented:
Don't be scared by some old commits, I kept the original and full history from the `sale-workflow` repository.

guewen commented on sale_condition_template/sale_order.py:
```diff
@@ -0,0 +1,76 @@
+
+    def action_invoice_create(self, cr, uid, ids,
+                              grouped=False,
+                              states=['confirmed', 'done', 'exception'],
```
Pylint says with reason:

> sale_condition_template/sale_order.py:52: [W0102(dangerous-default-value), SaleOrder.action_invoice_create] Dangerous default value ['confirmed', 'done', 'exception'] as argument

How should we handle that as the same default value is defined in the `super` method in the keyword argument? If we send `None`, the `super` method will have `None` instead of `['confirmed', 'done', 'exception']`.
Ideally it should be fixed in the core by setting `None` by default and with this code at the start of the method:

```
if states is None:
    states = ['confirmed', 'done', 'exception']
```

Now, in this module, I see 2 options:

A. None as sentinel and default value

```
def action_invoice_create(self, cr, uid, ids,
                          grouped=False,
                          states=None,
                          date_invoice=False, context=None):
    if states is None:
        states = ['confirmed', 'done', 'exception']
    _super = super(SaleOrder, self)
    inv_id = _super.action_invoice_create(cr, uid, ids,
                                          grouped=grouped,
                                          states=states,
                                          date_invoice=date_invoice,
                                          context=context)
```

B. Do not propagate the argument if not provided

```
def action_invoice_create(self, cr, uid, ids,
                          grouped=False,
                          states=None,
                          date_invoice=False, context=None):
    _super = super(SaleOrder, self)
    if states is None:
        inv_id = _super.action_invoice_create(cr, uid, ids,
                                              grouped=grouped,
                                              date_invoice=date_invoice,
                                              context=context)
    else:
        inv_id = _super.action_invoice_create(cr, uid, ids,
                                              grouped=grouped,
                                              states=states,
                                              date_invoice=date_invoice,
                                              context=context)
```

The option B. seems farfetched but it seems better to me because it does not duplicate the default value `['confirmed', 'done', 'exception']`.

yvaucher replied:
Maybe we could add a dict for kwargs to use the correct arguments for option B

guewen replied:
Adopted the @yvaucher's proposal, thanks

guewen commented:
It still misses the report part.

Commit:
Do not use a mutable default value.

It was there to stay compatible with the super method which has this
mutable list as default value. This fix removes it and now, the argument
is passed to the super method only if it is defined, so the super method
correctly set the default value.

Commit:
Display the note fields on the report

guewen commented:
I just added the fields on the report.

yvaucher commented on sale_condition_template/README.rst:
```diff
@@ -0,0 +1,28 @@
+Sale Conditions
+===============
+
+Adds template condition texts on sales orders and invoices.
```
Description to change, this modules doesn't add template on invoices here is a proposal:

```
Complete module `Invoice conditions` by adding template condition texts
on sales orders.

Templates are categorized by their position on the document.

Two positions are available:

- before sale order lines
- after sale order lines

Texts are copied on the invoice when you will create invoice from sale order.
```

yvaucher commented:
apart from description :+1:

jgrandguillaume commented:
Hi,

Thanks for the work ! What I saw:
- We need a way to manage (create and delete) the sales conditions. Can you add a menu in sales configuration ?
- In SO form, it become difficult to see everything. Can you put all top and bottom condition in a new tab ?

Otherwise, it's a great job.

Regards,

Joël

yvaucher commented:
@jgrandguillaume I wanted to make a "what you see is what you get".

Previously we had all notes at the same place as the additional informations. Plus templates m2o on an other tab.

I agree that with big condition your lines will step on the bottom. However those conditions shouldn't take more place than we have on the invoice page.
In this new tab we may need to add some more informative text to not let the user lost.

An other way I could think of is to collapse the condition text and uncollapse when editing it to be more readable but not sure we can do this in an easy way.

guewen commented:
Thanks for your review @jgrandguillaume,
The menu is already there: https://github.com/OCA/sale-reporting/pull/16/files#diff-253f4922b5fbca2cd12410ab96c5cd72R4

I agree that the text fields take too much place and we should find a better way to display them.

Your review on this one would be nice too because it is a pre-requisite: https://github.com/OCA/account-invoice-reporting/pull/15

gurneyalex commented:
:+1: waiting for travis to say "green"!
